In [6]:
import pandas as pd
import numpy as np
import matplotlib.pyplot as plt
%matplotlib inline
import pickle

In [1]:
import pickle

enron_data = pickle.load(open("../final_project/final_project_dataset.pkl", "rb"))

In [16]:
df = pd.DataFrame(enron_data).transpose()

#Setting proper NaN values
df = df.replace('NaN',np.NaN)

In [17]:
df.head()

bonus deferral_payments deferred_income director_fees  \
ALLEN PHILLIP K     4175000           2869717        -3081055           NaN   
BADUM JAMES P           NaN            178980             NaN           NaN   
BANNANTINE JAMES M      NaN               NaN           -5104           NaN   
BAXTER JOHN C       1200000           1295738        -1386055           NaN   
BAY FRANKLIN R       400000            260455         -201641           NaN   

                                 email_address exercised_stock_options  \
ALLEN PHILLIP K        phillip.allen@enron.com                 1729541   
BADUM JAMES P                              NaN                  257817   
BANNANTINE JAMES M  james.bannantine@enron.com                 4046157   
BAXTER JOHN C                              NaN                 6680544   
BAY FRANKLIN R             frank.bay@enron.com                     NaN   

                   expenses from_messages from_poi_to_this_person  \
ALLEN PHILLIP K       13868          2195                      47   
BADUM JAMES P          3486           NaN                     NaN   
BANNANTINE JAMES M    56301            29                      39   
BAXTER JOHN C         11200           NaN                     NaN   
BAY FRANKLIN R       129142           NaN                     NaN   

                   from_this_person_to_poi        ...         \
ALLEN PHILLIP K                         65        ...          
BADUM JAMES P                          NaN        ...          
BANNANTINE JAMES M                       0        ...          
BAXTER JOHN C                          NaN        ...          
BAY FRANKLIN R                         NaN        ...          

                   long_term_incentive    other    poi restricted_stock  \
ALLEN PHILLIP K                 304805      152  False           126027   
BADUM JAMES P                      NaN      NaN  False              NaN   
BANNANTINE JAMES M                 NaN   864523  False          1757552   
BAXTER JOHN C                  1586055  2660303  False          3942714   
BAY FRANKLIN R                     NaN       69  False           145796   

                   restricted_stock_deferred  salary shared_receipt_with_poi  \
ALLEN PHILLIP K                      -126027  201955                    1407   
BADUM JAMES P                            NaN     NaN                     NaN   
BANNANTINE JAMES M                   -560222     477                     465   
BAXTER JOHN C                            NaN  267102                     NaN   
BAY FRANKLIN R                        -82782  239671                     NaN   

                   to_messages total_payments total_stock_value  
ALLEN PHILLIP K           2902        4484442           1729541  
BADUM JAMES P              NaN         182466            257817  
BANNANTINE JAMES M         566         916197           5243487  
BAXTER JOHN C              NaN        5634343          10623258  
BAY FRANKLIN R             NaN         827696             63014  

[5 rows x 21 columns]

In [21]:
df.info()

<class 'pandas.core.frame.DataFrame'>
Index: 146 entries, ALLEN PHILLIP K to YEAP SOON
Data columns (total 21 columns):
bonus                        146 non-null object
deferral_payments            146 non-null object
deferred_income              146 non-null object
director_fees                146 non-null object
email_address                146 non-null object
exercised_stock_options      146 non-null object
expenses                     146 non-null object
from_messages                146 non-null object
from_poi_to_this_person      146 non-null object
from_this_person_to_poi      146 non-null object
loan_advances                146 non-null object
long_term_incentive          146 non-null object
other                        146 non-null object
poi                          146 non-null object
restricted_stock             146 non-null object
restricted_stock_deferred    146 non-null object
salary                       146 non-null object
shared_receipt_with_poi      146 non-null objec

In [18]:
df.describe()

bonus deferral_payments deferred_income director_fees email_address  \
count    146               146             146           146           146   
unique    42                40              45            18           112   
top      NaN               NaN             NaN           NaN           NaN   
freq      64               107              97           129            35   

       exercised_stock_options expenses from_messages from_poi_to_this_person  \
count                      146      146           146                     146   
unique                     102       95            65                      58   
top                        NaN      NaN           NaN                     NaN   
freq                        44       51            60                      60   

       from_this_person_to_poi        ...        long_term_incentive other  \
count                      146        ...                        146   146   
unique                      42        ...                         53    93   
top                        NaN        ...                        NaN   NaN   
freq                        60        ...                         80    53   

          poi restricted_stock restricted_stock_deferred salary  \
count     146              146                       146    146   
unique      2               98                        19     95   
top     False              NaN                       NaN    NaN   
freq      128               36                       128     51   

       shared_receipt_with_poi to_messages total_payments total_stock_value  
count                      146         146            146               146  
unique                      84          87            126               125  
top                        NaN         NaN            NaN               NaN  
freq                        60          60             21                20  

[4 rows x 21 columns]

#### 13. Quiz: size of the Enron Dataset
How many datapoints (e.g. people) are in the dataset?

In [19]:
len(df)

146

There are 146 datapoints in the dataset.

#### 14. Quiz: Features in the Enron Dataset

For each person, how many features are available?

In [20]:
len(df.columns)

21

There are 21 features in the dataset.

#### 15. Quiz: Finding POIs in teh Enron Data
The “poi” feature records whether the person is a person of interest, according to our definition. How many POIs are there in the E+F dataset?

In other words, count the number of entries in the dictionary where
data[person_name]["poi"]==1

In [23]:
len(df[df['poi'] == 1])

18

There are 18 persons of interest in the dataset.

#### 16. Quiz: How many POIs exist?
We compiled a list of all POI names (in ../final_project/poi_names.txt) and associated email addresses (in ../final_project/poi_email_addresses.py).

How many POI’s were there total? (Use the names file, not the email addresses, since many folks have more than one address and a few didn’t work for Enron, so we don’t have their emails.)

In [55]:
x = 2

if x == 1:
    print('cat')
elif x == 2:
    print('hah')

hah


In [80]:
#Loading file
df_poi_names = pd.read_csv('../final_project/poi_names.txt', skiprows=2, header=None, delim_whitespace=True, names = ['have_inbox', 'lastname', 'firstname'])

#Transforming df: removing trailing comma in last names
df_poi_names['lastname'] = df_poi_names['lastname'].map(lambda x: x[:-1])

#Trandforming df: Making poi column boolean (True/False)
df_poi_names['have_inbox'] = df_poi_names['have_inbox'].map(lambda x: True if x == '(y)' else (False if x == '(n)' else print('ERROR VALUE:', x)) )

In [79]:
len(df_poi_names)

35

There were 35 POIs total.

In [78]:
# import sys
# sys.path.append('../final_project/')

# from  poi_email_addresses import poiEmails

# len(poiEmails())

90

#### 17. Quiz: Problems With Incomplete Data

As you can see, we have many of the POIs in our E+F dataset, but not all of them. Why is that a potential problem?

We will return to this later to explain how a POI could end up not being in the Enron E+F dataset, so you fully understand the issue before moving on.

Answer:
Among many other reasons this is a problem because with almost half (17 of 35) of the POIs missing from the dataset it will be hard to conclude anything, as each person might have a different role in the scandal.  

Video answer:
"There are a few things you could say here, but our main thought is about having enough data to really learn the patterns.  In general, more data is always better--only having 18 data points doesn't give you that many examples to learn from."



#### 18. Quiz: Query the dataset 1

Like any dict of dicts, individual people/features can be accessed like so:

enron_data["LASTNAME FIRSTNAME"]["feature_name"]
or, sometimes 
enron_data["LASTNAME FIRSTNAME MIDDLEINITIAL"]["feature_name"]

What is the total value of the stock belonging to James Prentice?

In [86]:
df.loc['PRENTICE JAMES','total_stock_value']

1095040

The total value of stock belonging to James Prentice is 1 095 040.

#### 19. Quiz: Query the dataset 2

Like any dict of dicts, individual people/features can be accessed like so:

enron_data["LASTNAME FIRSTNAME"]["feature_name"]

How many email messages do we have from Wesley Colwell to persons of interest?

In [89]:
df.loc['COLWELL WESLEY',:]

bonus                                      1200000
deferral_payments                            27610
deferred_income                            -144062
director_fees                                  NaN
email_address                wes.colwell@enron.com
exercised_stock_options                        NaN
expenses                                     16514
from_messages                                   40
from_poi_to_this_person                        240
from_this_person_to_poi                         11
loan_advances                                  NaN
long_term_incentive                            NaN
other                                       101740
poi                                           True
restricted_stock                            698242
restricted_stock_deferred                      NaN
salary                                      288542
shared_receipt_with_poi                       1132
to_messages                                   1758
total_payments                 

We have 11 email messages from Wesley Colwell to POIs.

#### 20. Quiz: Query the dataset 3

Like any dict of dicts, individual people/features can be accessed like so:

enron_data["LASTNAME FIRSTNAME"]["feature_name"]

or

enron_data["LASTNAME FIRSTNAME MIDDLEINITIAL"]["feature_name"]

What’s the value of stock options exercised by Jeffrey K Skilling?

In [93]:
# df.loc['SKILLING K. JEFFREY',:]

df.filter(like='SKILLING', axis=0)['exercised_stock_options']

SKILLING JEFFREY K    19250000
Name: exercised_stock_options, dtype: object

Jeffrey K Skilling exercised stock options for a value of 19 250 000.

21. Quiz: Research the Enron Fraud

Which of these schemes was Enron not involved in?

- selling assets to shell companies at the end of each month, and buying them back at the beginning of the next month to hide accounting losses
- causing electrical grid failures in California
- illegally obtained a government report that enabled them to corner the market on frozen concentrated orange juice futures
- conspiring to give a Saudi prince expedited American citizenship
- a plan in collaboration with Blockbuster movies to stream movies over the internet


Answer:


#### 25. Quiz: Follow the Money

Of Lay, Skilling and Fastow, who took home the most money? How much money was it?

In [108]:
df['total_payments'].sort_values(ascii)

TOTAL                      309886585.0
LAY KENNETH L              103559793.0
FREVERT MARK A              17252530.0
BHATNAGAR SANJAY            15456290.0
LAVORATO JOHN J             10425757.0
SKILLING JEFFREY K           8682716.0
MARTIN AMANDA K              8407016.0
BAXTER JOHN C                5634343.0
BELDEN TIMOTHY N             5501630.0
DELAINEY DAVID W             4747979.0
WHALLEY LAWRENCE G           4677574.0
ALLEN PHILLIP K              4484442.0
SHERRIFF JOHN R              4335388.0
MCMAHON JEFFREY              4099771.0
HAEDICKE MARK E              3859065.0
FALLON JAMES B               3676340.0
KITCHEN LOUISE               3471141.0
MULLER MARK S                3202070.0
HORTON STANLEY C             3131860.0
PAI LOU L                    3123383.0
HUMPHREY GENE E              3100224.0
SHANKMAN JEFFREY A           3038702.0
ECHOLS JOHN B                2692324.0
BOWEN JR RAYMOND M           2669589.0
LEFF DANIEL P                2664228.0
KOPPER MICHAEL J         

Lay took home the most money, with a total of 103 559 793.

#### Quiz 27: Dealing with Unfilled Features

How many folks in this dataset have a quantified salary? What about a known email address?

In [113]:
print("Has salary data:", len(df[df['salary'].isna() == False]))
print("Has email address:", len(df[df['email_address'].isna() == False]))

Has salary data: 95
Has email address: 111


#### Quiz 29: Missing POIs 1

As you saw a little while ago, not every POI has an entry in the dataset (e.g. Michael Krautz). That’s because the dataset was created using the financial data you can find in final_project/enron61702insiderpay.pdf, which is missing some POI’s (those absences propagated through to the final dataset). On the other hand, for many of these “missing” POI’s, we do have emails.

While it would be straightforward to add these POI’s and their email information to the E+F dataset, and just put “NaN” for their financial information, this could introduce a subtle problem. You will walk through that here.

How many people in the E+F dataset (as it currently exists) have “NaN” for their total payments? What percentage of people in the dataset as a whole is this?

In [143]:
count_nan_totalpay = len(df[df['total_payments'].isna()])

percentage_nan_totalpay = round(count_nan_totalpay / len(df) * 100,2)

print("Count:", count_nan_totalpay)
print("Percentage of total:", percentage_nan_totalpay)

Count: 21
Percentage of total: 14.38


#### Quiz 30: Missing POIs 2

How many POIs in the E+F dataset have “NaN” for their total payments? What percentage of POI’s as a whole is this?

In [144]:
df_pois = df[df['poi'] == 1]

count_nan_totalpay = len(df_pois[df_pois['total_payments'].isna()])

percentage_nan_totalpay = round(count_nan_totalpay / len(df_pois) * 100,2)

print("Count:", count_nan_totalpay)
print("Percentage of total:", percentage_nan_totalpay)

Count: 0
Percentage of total: 0.0


#### Quiz 31: Missing POIs 3

If a machine learning algorithm were to use total_payments as a feature, would you expect it to associate a “NaN” value with POIs or non-POIs?

I would expect it to assoicate a NaN value with non-POIs

#### Quiz 31: Missing POIs 4

If you added in, say, 10 more data points which were all POI’s, and put “NaN” for the total payments for those folks, the numbers you just calculated would change.
What is the new number of people of the dataset? What is the new number of folks with “NaN” for total payments?

In [147]:
print("New people count:", len(df) + 10)
print("New NaN count:", 31)

New people count: 156
New NaN count: 31


#### Quiz 31: Missing POIs 5

What is the new number of POI’s in the dataset? What is the new number of POI’s with NaN for total_payments?

In [148]:
print("New POI count:", len(df_pois) + 10)
print("New NaN POI count:", 10)

New POI count: 28
New NaN POI count: 10


#### Quiz 31: Missing POIs 6


Once the new data points are added, do you think a supervised classification algorithm might interpret “NaN” for total_payments as a clue that someone is a POI?

Answer: 
Yes, as the percentage changed I think it would skew NaN total_payments value towards being a POI.